File: mtp.cpp
#include <bits/stdc++.h>
using namespace std;

int main()
{
	unordered_map<string,int>symptoms;
	vector<vector<string>> patients;
	vector<string> data;
	string s;
	ifstream file("patientjourneyfinder.csv");
	while(getline(file,s)){
		data.push_back(s);
	}
	for(string i:data){
		string str="";
		vector<string>temp;
		for(int j=0;j<i.length();j++){
			if(i[j]==','){
				temp.push_back(str);
				str="";
				continue;
			}
			str+=i[j];
		}
		patients.push_back(temp);
	}
	patients.erase(patients.begin());
	ifstream sy("symptoms.csv");
	s="";
	vector<string>sysrare;
	while(getline(sy,s)){
		sysrare.push_back(s);
	}
	sysrare.erase(sysrare.begin());
	for(string i:sysrare)if(symptoms.find(i)==symptoms.end())symptoms[i]=1;
	//storing the physician name across every patients.
	unordered_map<string,string> hashedphysician;
	vector<vector<string>> hashedpatient;
	vector<pair<int,string>> patientNameLikelihood;
	for(int i=0;i<patients.size();i++){
		vector<string> temp;
		temp.push_back(patients[i][0]);
		hashedphysician[patients[i][0]]=patients[i][12];
		for(int j=13;j<=27;j++)temp.push_back(patients[i][j]);
		hashedpatient.push_back(temp);
	}
	/**
	* calculating likelihood
	*
	*/
	for(int i=0;i<hashedpatient.size();i++){
		string name=hashedpatient[i][0];
		int count=0;
		for(int j=1;j<hashedpatient[i].size();j++){
			if(symptoms[hashedpatient[i][j]]){
				count++;
			}
		}
		patientNameLikelihood.push_back(make_pair(count,name));
	}
	sort(patientNameLikelihood.begin(),patientNameLikelihood.end());
	reverse(patientNameLikelihood.begin(),patientNameLikelihood.end());
	//
	// through physician
	unordered_map<string,int> doctor;
	unordered_map<string,int> doctorcount;
	for(int i=0;i<patientNameLikelihood.size();i++){
		if(doctor.find(hashedphysician[patientNameLikelihood[i].second])==doctor.end()){
			doctor[hashedphysician[patientNameLikelihood[i].second]]=patientNameLikelihood[i].first;
			doctorcount[hashedphysician[patientNameLikelihood[i].second]]=1;
		}
		else{
			doctorcount[hashedphysician[patientNameLikelihood[i].second]]=1;
		 	doctor[hashedphysician[patientNameLikelihood[i].second]]+=patientNameLikelihood[i].first;
		 }
	}

	/*likelihod through physician side*/
	vector<pair<double,string>>doc;
	auto it=doctor.begin();
	auto iit=doctorcount.begin();
	while(it!=doctor.end()){
		doc.push_back(make_pair((double)it->second/(double)iit->second,it->first));
		it++;
		iit++;
	}

	sort(doc.begin(),doc.end());
	reverse(doc.begin(),doc.end());
	/*List through patients side*/
	vector<string> physician;
	for(int i=0;i<patientNameLikelihood.size();i++){
		if(patientNameLikelihood[i].first>0){
			physician.push_back(hashedphysician[patientNameLikelihood[i].second]);
		}
	}
	/*List through doctor side*/
	vector<string>doctorname;
	for(int i=0;i<doc.size();i++){
		if(doc[i].first>0)
			doctorname.push_back(doc[i].second);
	}

	// for(string i:doctorname)cout<<i<<endl;                                     ----------------------------------------------------------------
	//physicianslist is a list of a physician
	unordered_map<string,bool> phy;
	int number=0;
	vector<string> physicianslist;
	cout<<"\nPhysician Id's\n\n";
	for(int i=0;i<physician.size();i++){
		if(phy.find(physician[i])==phy.end()){
			physicianslist.push_back(physician[i]);
			number++;
			phy[physician[i]]=true;
		}
	}
	for(string i:doctorname){
		if(phy.find(i)==phy.end()){
			physicianslist.push_back(i);
			number++;
			phy[i]=true;
		}
	}
	/*
	doctor details
	hard code right now;

	*/
	vector<vector<string>>doctorDetail;
	vector<string>tempDocDetail;
	ifstream doctorlist("doctorlist.csv");
	string s1="";
	while(getline(doctorlist,s1)){
		tempDocDetail.push_back(s1);
	}
	for(string i:tempDocDetail){
		string str="";
		vector<string>temp;
		for(int j=0;j<i.length();j++){
			if(i[j]=='\t'){
				temp.push_back(str);
				str="";
				continue;
			}
			str+=i[j];
		}
		temp.push_back(str);
		doctorDetail.push_back(temp);
	}
	unordered_map<string,vector<string>> hasheddoctordetail;
	for(int i=0;i<doctorDetail.size();i++){
		for(int j=1;j<doctorDetail[i].size();j++){
			hasheddoctordetail[doctorDetail[i][0]].push_back(doctorDetail[i][j]);
		}
	}
		/*

		*/
	for(string i:physicianslist){
		cout<<i<<endl;
		vector<string> details;
		details=hasheddoctordetail[i];
		for(string j:details)cout<<j<<"\t";
		cout<<endl<<endl;
	}
	cout<<"\nNumber of physician treating rare disease:\t"<<number<<endl;
	return 0;
}

File: mtp.py
import csv
import time
t1=time.time()
patients = []
with open('patientjourneyfinder.csv','r') as f: 
	reader = csv.reader(f)
	for row in reader: patients.append(row)
patients = patients[1:]
# print len(patients)
sysrare = []
with open('symptoms.csv','r') as f:
	reader = csv.reader(f) 
	for row in reader: sysrare.append(row[0])
symptoms = { x:1 for x in set(sysrare[1:]) }

hashedpatients,hashedphysician = [([x[0]]+x[13:28]) for x in patients], { x[0]:x[12] for x in patients }
patientNameLikelihood = (sorted([(sum([1 for x in patient[1:] if x in symptoms]),patient[0]) for patient in hashedpatients]))[::-1]

doctors,doctorCount = {},{}
for patient in patientNameLikelihood:
	if hashedphysician[patient[1]] in doctors:
		doctors[hashedphysician[patient[1]]] += patient[0]
		doctorCount[hashedphysician[patient[1]]] += 1
	else:
		doctors[hashedphysician[patient[1]]] = patient[0]
		doctorCount[hashedphysician[patient[1]]] = 1

doc = []
for k,v in doctors.iteritems(): doc.append((float(doctors[k])/float(doctorCount[k]),k))
doc = (sorted(doc))[::-1]

physicians,doctorName = [hashedphysician[x[1]] for x in patientNameLikelihood if x[0]>0],[x[1] for x in doc if x[0]>0]
phy,number,physiciansList = {},0,[]

for physician in physicians:
	if not(physician in phy):
		physiciansList.append(physician)
		phy[physician] = True
		number += 1
for doctor in doctorName:
	if not(doctor in phy):
		number += 1
		physiciansList.append(doctor)
		phy[doctor] = True

doctorDetail = []
with open("doctorlist.csv",'r') as f:
	reader = csv.reader(f)
	for row in reader: doctorDetail.append(row[0].split('\t'))

hashedDoctorDetail = {}
for d in doctorDetail: hashedDoctorDetail[d[0]] = d[1:]
t2=time.time()
for i in physiciansList:
	print i
	print "\t".join(hashedDoctorDetail[i])

print "\nNumber of physician treating rare disease:",number

print t2-t1
